### ISSUE TITLE
Reduce network spam from mouse movements in games

### ISSUE THREAD

**Issue by cdsmith** [google/codeworld#379]
Games generate a LOT of mouse movement events, which are each propagated to all clients.  We should be able to reduce this.  For example, suppose each client tracked when it already had a mouse movement sent within the last 50 milliseconds; and if so, coalesced all consecutive mouse events and postponed broadcasting them until 50 milliseconds had passed.

We could also look into a trick similar to the redraw-avoidance.  It's less straight-forward, though, because  even if the MouseMovement doesn't modify local states, it still needs to be sent because it might matter in whatever (as yet unknown) state the game will be in, in the merged canonical event stream.  So that's not great.  BUT, if we could introspect that MouseMovement events are _always_ the identity function for a given game, then we could save a lot of traffic.

It's not immediately clear to me how to discover this.  @nomeata knows a fair bit about this: any ideas?

**Comment by cdsmith** [comment]
What's tripping me up about the introspection approach is this:  The naive approach...

    let undefState = undefined
    undefName <- makeStableName undefState
    let newState = event playerID (MouseMovement undefined) undefState

looks approximately right.  But what now?  Of course newState isn't the same as undefState, because it's an unevaluated thunk.  But if I evaluate it, it will almost surely crash.  It's like I want to evaluate it a *little* bit, to see if I can get back undefName one layer deep...

**Comment by nomeata** [comment]
>  It's less straight-forward, though, because even if the MouseMovement doesn't modify local states, it still needs to be sent because it might matter in whatever (as yet unknown) state the game will be in, in the merged canonical event stream.

I don’t think this is true. If the user moves the mouse during a non-committed state that is visible, and in that visible state the event does not affect the state, then clearly the intention was for the movement (or any event) to have an effect. And if we drop the event without doing anything with it (i.e. do not add it to the local event queue), consistency between the clients is still given.

So for any event, just check if it has an effect on the current state, and drop it if there is no effect.

**Comment by cdsmith** [comment]
I've been thinking a lot about this, and going back and forth.  I don't think there is any kind of a formal sense in which we can say that these movements aren't intended to have an effect.  There are plenty of cases in games where I may do something in anticipation of its having an effect.

On the other hand, maybe getting semantics right in a formal sense doesn't matter so much, and mouse movement events have a high cost-benefit ratio because they so rarely correspond to an intended action, and happen at such a high rate.  So the priority is definitely just doing what's necessary to keep the states aligned.  And the good news is that we can try almost anything, and the prediction stuff you wrote guarantees that we keep eventual consistency!

**Comment by nomeata** [comment]
> There are plenty of cases in games where I may do something in anticipation of its having an effect.

But that would make only sense if the user starts to anticipating the network lag, right?

**Comment by cdsmith** [comment]
Well, yes, and that a user's reaction time overlaps with the network lag.  That said: this is the kind of thing that might matter for competitive gaming, but not for CodeWorld games.  As long as we aren't skipping MousePress and KeyPress on this same theory, then I think I agree.

**Comment by nomeata** [comment]
>  As long as we aren't skipping MousePress and KeyPress on this same theory, then I think I agree.

I might even go that far!

Behaviour under your proposal: User presses button. Nothing happens. A moment later, new events from the server come in. The state updates. Suddenly the press had an effect. Result: The user noticed that the button did not do anything, and then later did turned out it did.

Behaviour under your proposal: User presses button. Nothing happens. A moment later, new events from the server come in. The state updates. Still nothing happened because of the event. Result: The button press did not have an effect _as it was right in the moment the button was pressed_. The behaviour is as expected, and it is almost as if this event (which does not do anything!) was handled instantaneously on all clients. :-)

Well, we’ll see how it goes in practice. It probably matters little.

**Comment by cdsmith** [comment]
Okay, let's find out.  We're still in the experimental stage, so I applied this optimization to all events.  As we're playing around with it, we'll discover if it leads to counter-intuitive results!

**Comment by cdsmith** [comment]
Closing.  If this causes problems, I can file the bug for it then.

**Comment by cdsmith** [comment]
Oh, I see.  I wanted this still open for coalescing behavior.  I'll re-open then.
